flyteplugins/go/tasks/plugins/webapi/agent/plugin.go

```diff
@@ -79,62 +83,120 @@ func (p Plugin) Create(ctx context.Context, taskCtx webapi.TaskExecutionContextR
-	}
+	taskType := admin.TaskType{Name: taskTemplate.Type, Version: taskTemplate.TaskTypeVersion}
+	agent, isSync := getFinalAgent(&taskType, p.cfg, p.agentRegistry)
 
 	finalCtx, cancel := getFinalContext(ctx, "CreateTask", agent)
 	defer cancel()
 
 	taskExecutionMetadata := buildTaskExecutionMetadata(taskCtx.TaskExecutionMetadata())
-	res, err := client.CreateTask(finalCtx, &admin.CreateTaskRequest{Inputs: inputs, Template: taskTemplate, OutputPrefix: outputPrefix, TaskExecutionMetadata: &taskExecutionMetadata})
-	if err != nil {
-		return nil, nil, err
-	}
 
-	// Restore unrendered template for subsequent renders.
-	if taskTemplate.GetContainer() != nil {
-		taskTemplate.GetContainer().Args = argTemplate
+	if isSync {
+		client, err := p.getSyncAgentClient(ctx, agent)
+		if err != nil {
+			return nil, nil, err
+		}
+		header := &admin.CreateRequestHeader{Template: taskTemplate, OutputPrefix: outputPrefix, TaskExecutionMetadata: &taskExecutionMetadata}
+		return p.ExecuteTaskSync(finalCtx, client, header, inputs)
 	}
 
-	// If the agent returned a resource, we assume this is a synchronous task.
-	// The state should be a terminal state, for example, SUCCEEDED, PERMANENT_FAILURE, or RETRYABLE_FAILURE.
-	if res.GetResource() != nil {
-		logger.Infof(ctx, "Agent is executing a synchronous task.")
-		return nil,
-			ResourceWrapper{
-				Phase:    res.GetResource().Phase,
-				State:    res.GetResource().State,
-				Outputs:  res.GetResource().Outputs,
-				Message:  res.GetResource().Message,
-				LogLinks: res.GetResource().LogLinks,
-			}, nil
+	// Use async agent client
+	client, err := p.getAsyncAgentClient(ctx, agent)
+	if err != nil {
+		return nil, nil, err
+	}
+	request := &admin.CreateTaskRequest{Inputs: inputs, Template: taskTemplate, OutputPrefix: outputPrefix, TaskExecutionMetadata: &taskExecutionMetadata}
+	res, err := client.CreateTask(finalCtx, request)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	logger.Infof(ctx, "Agent is executing an asynchronous task.")
 	return ResourceMetaWrapper{
 		OutputPrefix:      outputPrefix,
 		AgentResourceMeta: res.GetResourceMeta(),
-		Token:             "",
-		TaskType:          taskTemplate.Type,
+		TaskType:          taskType,
 	}, nil, nil
 }
 
+func (p Plugin) ExecuteTaskSync(
+	ctx context.Context,
+	client service.SyncAgentServiceClient,
+	header *admin.CreateRequestHeader,
+	inputs *flyteIdl.LiteralMap,
+) (webapi.ResourceMeta, webapi.Resource, error) {
+	stream, err := client.ExecuteTaskSync(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	headerProto := &admin.ExecuteTaskSyncRequest{
+		Part: &admin.ExecuteTaskSyncRequest_Header{
+			Header: header,
+		},
+	}
+
+	err = stream.Send(headerProto)
+	if err == nil {
+		inputsProto := &admin.ExecuteTaskSyncRequest{
+			Part: &admin.ExecuteTaskSyncRequest_Inputs{
+				Inputs: inputs,
+			},
+		}
+		err = stream.Send(inputsProto)
+		if err != nil {
+			return nil, nil, err
```

Comment:
nit, wrap errors so we know where things failed...
`fmt.Errorf("failed to send inputsProto. Error: %w", err)`